platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassPathProvider.kt

```diff
@@ -64,34 +67,58 @@ fun kotlinScriptClassPathProviderOf(project: Project) =
     project.serviceOf<KotlinScriptClassPathProvider>()
 
 
-internal
-typealias JarsProvider = () -> Collection<File>
-
-
 @ServiceScope(Scope.Build::class)
 class KotlinScriptClassPathProvider(
-    private val moduleRegistry: ModuleRegistry,
-    private val classPathRegistry: ClassPathRegistry,
-    private val coreAndPluginsScope: ClassLoaderScope,
-    private val gradleApiJarsProvider: JarsProvider,
+    gradleProperties: GradleProperties,
+    moduleRegistry: ModuleRegistry,
+    dependencyFactory: DependencyFactory,
+    classPathRegistry: ClassPathRegistry,
+    coreAndPluginsScope: ClassLoaderScope,
 ) : GradleApiClasspathProvider {
 
-    /**
-     * Generated Gradle API jar plus supporting libraries such as groovy-all.jar and generated API extensions.
-     */
     private
-    val gradleKotlinDslClasspath: ClassPath by lazy {
-        gradleApi + gradleApiExtensions + gradleKotlinDslJars
+    val gradleAbiClasspath: ClassPath by lazy {
+        moduleRegistry.findModule("gradle-public-api-legacy")
+            ?.allRequiredModulesClasspath
+            ?.asFiles
+            ?.let { files -> DefaultClassPath.of(files) }
```

Comment:
allRequiredModulesClasspath is already a `ClassPath`, why do we have this logic?